src/crimson/os/seastore/extentmap_manager.h

```diff
@@ -0,0 +1,119 @@
+#include "crimson/os/seastore/transaction_manager.h"
+
+#define PAGE_SIZE 4096
+#define EXTMAP_BLOCK_SIZE 4096
+
+namespace crimson::os::seastore {
+
+struct lext_map_val_t {
+  laddr_t laddr;
+  extent_len_t length = 0;
+
+  lext_map_val_t(
+    laddr_t laddr,
+    extent_len_t length)
+    : laddr(laddr), length(length) {}
+
+};
+class Extent
+{
+public:
+  objaddr_t logical_offset = 0;  //offset in object
+  laddr_t laddr;     // lextent start address aligned with block size.
+  extent_len_t length = 0;
+  explicit Extent(objaddr_t lo) : logical_offset(lo) { }
+
+  Extent(
+    objaddr_t lo,
+    laddr_t laddr,
+    extent_len_t length)
+    : logical_offset(lo), laddr(laddr), length(length) {}
+
+  void dump(ceph::Formatter* f) const;
+    friend bool operator<(const Extent &a, const Extent &b) {
+      return a.logical_offset < b.logical_offset;
+    }
+    friend bool operator>(const Extent &a, const Extent &b) {
+      return a.logical_offset > b.logical_offset;
+    }
+    friend bool operator==(const Extent &a, const Extent &b) {
+      return a.logical_offset == b.logical_offset;
+    }
+
+  ~Extent() {}
+};
+
+using ExtentRef = std::unique_ptr<Extent>;
+using extent_map_list_t = std::list<ExtentRef>;
+std::ostream &operator<<(std::ostream &out, const Extent &rhs);
+std::ostream &operator<<(std::ostream &out, const extent_map_list_t &rhs);
+
+struct extmap_root_t {
+  depth_t depth = 0;
+  laddr_t extmap_root_laddr;
+  extmap_root_t(depth_t dep, laddr_t laddr)
+  : depth(dep),
+    extmap_root_laddr(laddr) {}
+};
+
+using extmap_root_ref = std::shared_ptr<extmap_root_t>;
+
+/**
+ * Abstract interface for managing the object inner offset to logical addr mapping
+ * each onode has an extentmap tree for a particular onode.
+ */
+class ExtentMapManager {
+public:
+  using alloc_extmap_root_ertr = TransactionManager::alloc_extent_ertr;
+  using alloc_extmap_root_ret = alloc_extmap_root_ertr::future<>;
+  virtual alloc_extmap_root_ret alloc_extmap_root(Transaction &t) = 0;
+
+  //find the extents which overlap the object offset or after the offset
+  using find_lextent_ertr = TransactionManager::read_extent_ertr;
+  using find_lextent_ret = find_lextent_ertr::future<extent_map_list_t>;
+  virtual find_lextent_ret
+    find_lextent(Transaction &t, objaddr_t lo, extent_len_t len) = 0;
+
+  //add a new Extent
+  using add_lextent_ertr = TransactionManager::read_extent_ertr;
+  using add_lextent_ret = add_lextent_ertr::future<ExtentRef>;
+  virtual add_lextent_ret
+    add_lextent(Transaction &t, objaddr_t lo, lext_map_val_t val) = 0;
+
+  // remove an Extent
+  using rm_lextent_ertr = TransactionManager::read_extent_ertr;
+  using rm_lextent_ret = rm_lextent_ertr::future<bool>;
+  virtual rm_lextent_ret rm_lextent(Transaction &t, objaddr_t lo, lext_map_val_t val) = 0;
+
+  virtual ~ExtentMapManager() {}
+};
+using ExtentMapManagerRef = std::unique_ptr<ExtentMapManager>;
+
+namespace extentmap_manager {
```

Comment:
Explain the difference between these with comments as above.  For the first, for instance, the user is required to call alloc_extmap_root prior to use, right?